### DISCUSSION TITLE
handling ctrl + c in tests

### DISCUSSION THREAD

**Discussion by jakim929** [wagmi-dev/anvil.js#30] [Q&A] [answered]
Been using anviljs with vitest and am currently starting up a new chain `beforeEach` and stopping it `afterEach`. It's been great! 🔥 

One thing I noticed is that if I kill a test with `ctrl + c` the anvil process doesn't get killed. Then I have to go manually kill the process to restart the tests since it is using the same port. 

Is there some recommended way to setup tests to handle interrupts?

**Comment by fubhy** [comment, marked as answer]
Yeah, I also observed that in my first experiments with the test setup. Vitest doesn't cleanly / gracefully shutdown the worker threads when the main process is terminated. To my knowledge, there's no way for us to do anything about this sadly. If you want to spawn your instances in `beforeEach` and `afterEach` directly in the worker threads (without the global proxy), then you'll have to open an issue in the `vitest` repository to get this resolved. I'm not sure why it currently behaves like this and whether they'd be open to improving / changing that or not.

That said, we've since discovered that it's actually better to NOT spawn the processes in `beforeEach` and `afterEach` anyways and instead spawn them all in the global main process. Check out the examples in this repository to see how that works. You can still reset the individual instances in `beforeEach` or `afterEach` to your liking. But you get several benefits: E.g. you can share snapshots between tests and you can share the in-memory anvil cache between tests, etc. That approach has proven to be much more flexible and performant for our needs.
